### PR TITLE
Conversion function from ecp group to PSA curve

### LINKED ISSUE #7764: Conversion function from ecp group to PSA curve
There should be a simple way to convert from a `mbedtls_ecp_group` or `mbedtls_ecp_group_id` to the corresponding PSA encoding (curve and bit-size). This is needed when migrating code that needs low-level `ecp.h` code to create a key, which is then used through the PSA API. For example, to import a compressed point using `mbedtls_ecp_point_read_binary` followed by `mbedtls_ecp_point_write_binary` and `psa_import_key`.

Goals of this issue:

* Design the new API.
* Implement and test it.
* List it in the upcoming PSA transition guide.

### PR DESCRIPTION
**Pull request by valeriosetti**
Resolves #7764

Depends on #8668

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required
- [x] **tests** provided

### CODE REVIEW

File: library/psa_crypto_ecp.c
Comment (mpg) [inline library/psa_crypto_ecp.c]: Perhaps also tell static analyzers: `(void) mbedtls_ecc_group_to_psa(...)`. Coverity for instance tends to warn when the return value is ignored in one call while is usually isn't. (Edit: removed in a later commit.)

File: library/psa_util.c
Comment (mpg) [inline library/psa_util.c:265]: There was some ambiguity in the task description: removing `bits_is_sloppy` can be done either by behaving as if it's always true or always false. You chose the former. Can you explain why?

Also, this is something we want to document if we keep it that way. (Which is actually an argument against maintaining "sloppy" behaviour: public functions should be simple to document.)

Edit (after reading more of the code): actually, since there is only one place that potentially used `bits_is_sloppy == 1`, it would seem more natural to me for the function to always behave as false, and then for the only call site that needed sloppy behaviour to have a special case here. (Could be an explicit check for one of the two (family,size) pairs concerned, or perhaps we would include that in `check_ecc_parameters()`: for example instead of taking `int allow_bit_size_roundup` it could take `size_t *actual_bits` - if `NULL`, then checking is strict, if non-`NULL` checking is more relaxed and `*actual_bits` is updated upon return.

Wdyt?

File: tests/suites/test_suite_psa_crypto.data
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.data:7463]: I guess the CI will complain about a missing newline at the end of this file.

File: tests/suites/test_suite_psa_crypto.function
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function:9493]: This is only testing that the functions are inverse of each other, not that they are correct. The risk of errors here seems low, but as a matter of principle I'd like it better if the test functions took parameters like `exp_ecc_family` and `exp_bits` and checked that family and bits are as expected.

File: tests/suites/test_suite_psa_crypto.function
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function:9503]: You can actually use `MBEDTLS_ECP_DP_MAX` for that.

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good. Apart from minor things, the only non-trivial point is about removal of `bit_size_is_sloppy`.

**Comment by mpg** [comment]
Also, the CI has quite a lot of things to say. I didn't check, but since I saw a test case failing in the first component, I assume we'll have a lot of occurrences of the same failure:
```
[2023-12-28T19:38:55.711Z]          PSA import ECC_KEY_PAIR(SECP_K1) 224-bit curve not supported ...... FAILED
[2023-12-28T19:38:55.711Z]            psa_import_key(&attributes, key_material->x, key_material->len, &key_id) == PSA_ERROR_NOT_SUPPORTED
[2023-12-28T19:38:55.711Z]            at line 26, C:/Windows/workspace/mbed-tls-pr-head_PR-8664-head/worktrees/tmpixgncy05/tests/suites/test_suite_psa_crypto_not_supported.function
[2023-12-28T19:38:55.711Z]            lhs = 0xffffffffffffff79 = -135
[2023-12-28T19:38:55.711Z]            rhs = 0xffffffffffffff7a = -134
```
I suggest we fix this one first, then check if there were others hidden in the results.

**Comment by valeriosetti** [comment]
> Also, the CI has quite a lot of things to say. I didn't check, but since I saw a test case failing in the first component, I assume we'll have a lot of occurrences of the same failure:

Unfortunately yes. This is due to the fact that I tried to improve `mbedtls_psa_ecp_load_representation()` in order to return `PSA_ERROR_INVALID_ARGUMENT` and `PSA_ERROR_INVALID_ARGUMENT` when appropriate, instead of returning always the latter. I will try to check if there is a way to overcome this limitation. 

Side note: those tests are supposed to be skipped (according to their description), but they are executed because `SECP224K1` is commented out in `crypto_config.h`. Perhaps also this thing might be fixed sooner or later in this automatic generated test.

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

**Comment by mpg** [comment]
I had a look at the stability checks results, and found them OK:
- A function is renamed - but that function was marked as internal in the previous release.
- Some tests are marked as having disappeared but they just changed due to fixed the value for the secp224k1 private key.

File: scripts/mbedtls_dev/asymmetric_key_data.py
Comment (mpg) [inline scripts/mbedtls_dev/asymmetric_key_data.py:44]: Note: this might not be the correct fix after all, see https://github.com/Mbed-TLS/mbedtls/pull/4076#discussion_r1440275352

File: library/psa_crypto_ecp.c
Comment (gilles-peskine-arm) [inline library/psa_crypto_ecp.c:85]: Our PSA implementation [does not and will not support secp224k1](https://github.com/Mbed-TLS/mbedtls/issues/3541). So we could save a few bytes of code size and maybe a little bit of complexity by removing the existing support. (Keep a comment, since it would be somewhat surprising that secp224k1 is missing in the PSA code even though it's supported in the legacy API.)

For `check_ecc_parameters`, we just have to return `INVALID_ARGUMENT` or `NOT_SUPPORTED`. For `mbedtls_ecc_group_from_psa` and `mbedtls_ecc_group_to_psa`, we need to make sure the behavior is consistent with what we say about recognized curves in the documentation.

File: tests/suites/test_suite_psa_crypto.data
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:7463]: Please add a test case for null values:
```
ecc_conversion_functions:MBEDTLS_ECP_DP_NONE:0:0
```

Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.data:7463]: Please add a couple of negative cases for PSA→legacy: at least one invalid family and at least one valid family with a size that's valid in some other family but not that one. Also now would be a good time to add test cases with `PSA_ECC_FAMILY_TWISTED_EDWARDS` and valid sizes (255, 448), mapping to `MBEDTLS_ECP_DP_NONE` since these curves aren't supported yet.

For clarity it would be best to have separate test functions for the negative cases that just go one way.

File: docs/psa-transition.md
Comment (gilles-peskine-arm) [inline docs/psa-transition.md:746]: Most readers should use these conversion functions and won't need the table. So please move this text before the table. Something like:

> 
A curve is fully determined by a curve family identifier and the private key size in bits. You can use the following functions to convert between the PSA and legacy elliptic curve designations: (…)
>
> The following table gives the correspondence between legacy and PSA elliptic curve designations. (…)

File: docs/psa-transition.md
Comment (gilles-peskine-arm) [inline docs/psa-transition.md:747]: Please hyperlink the functions to their documentation, as is done for other functions in this document. You can find the URL in the [readthedocs build](https://mbed-tls--8664.org.readthedocs.build/projects/api/en/8664/), replacing `https://mbed-tls--8664.org.readthedocs.build/projects/api/en/8664/` by `https://mbed-tls.readthedocs.io/projects/api/en/development/`.

File: ChangeLog.d/7764.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/7764.txt:2]: Minor: we normally word those with an active present tense. “Add functions …”.

File: ChangeLog.d/7764.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/7764.txt:4]: Minor: “to convert from … and … and vice versa, respectively” → “to convert between … and …”

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
This mostly looks good, and fulfils the requirements for these functions in https://github.com/Mbed-TLS/mbedtls/pull/8657 as it currently stands. I'm requesting some minor improvements to the documentation and to the tests. Also an optional suggestion for a minor simplification in the implementation.

File: library/psa_crypto_ecp.c
Comment (valeriosetti) [reply library/psa_crypto_ecp.c:85]: I suspect that removing the support for secp224k1 might end up in the failure I mentioned in a [previous comment](https://github.com/Mbed-TLS/mbedtls/pull/8664#issuecomment-1871849371), in particular:
> those tests are supposed to be skipped (according to their description), but they are executed because SECP224K1 is commented out in crypto_config.h. Perhaps also this thing might be fixed sooner or later in this automatic generated test.

File: tests/suites/test_suite_psa_crypto.function
Comment (mpg) [inline tests/suites/test_suite_psa_crypto.function:9504]: I think we want to test both invalid (using `ECP_DP_MAX`) and `NONE`.

File: library/psa_crypto_ecp.c
Comment (mpg) [inline library/psa_crypto_ecp.c:85]: This seems to contradict the comment, which has 225.

File: ChangeLog.d/7764.txt
Comment (mpg) [inline ChangeLog.d/7764.txt:4]: I think "between ... and ..." already means it goes both way, so the "and vice versa" is superfluous (and the "respectively" as well.

File: library/psa_crypto_ecp.c
Comment (valeriosetti) [reply library/psa_crypto_ecp.c:85]: 225 is for the private key; I think public is still 224. Following "adjustments" I did for other families here I should change 232 to 225 for private keys, but since we're not supporting secp224k1 I didn't add this piece of code here. 
Btw I will fix the comment to be clearer.

**Review by mpg — COMMENTED** [review]
Note sure about handling or secp224k1 yet. Perhaps we should get https://github.com/Mbed-TLS/mbedtls/pull/8668/files merged first and then rebase this one on top?

File: ChangeLog.d/7764.txt
Comment (gilles-peskine-arm) [inline ChangeLog.d/7764.txt:4]: “Between X and Y and Z” looks weird: the natural reading is that there are conversions between any two of X, Y and Z. I think you don't need to mention the bit size here at all: it's just between `mbedtls_ecp_group_id` and `psa_ecc_family_t`, or between mbedtls and PSA curve identifiers.

File: include/mbedtls/psa_util.h
Comment (gilles-peskine-arm) [inline include/mbedtls/psa_util.h:124]: `config_adjust_legacy_from_psa.h` is an internal header, it won't help users. Legacy support is enabled iff `MBEDTLS_ECP_DP_xxx_ENABLED` is defined, but I'm not sure if we need to explain that here. Isn't it enough to just say

> If the curve is supported in the PSA API, this function succeeds, even if the curve is not supported by the ECP module. If the curve is not supported in the PSA API, this function returns 0.

Maybe we even don't need this note: we could just clarify it in the `return` statement above (replace “on success” by what causes success).

File: include/mbedtls/psa_util.h
Comment (gilles-peskine-arm) [inline include/mbedtls/psa_util.h:119]: This looks like the function returns 0 if the curve is not supported in the legacy API, but according to the note below, this is not true if the curve is supported in PSA.

File: include/mbedtls/psa_util.h
Comment (gilles-peskine-arm) [inline include/mbedtls/psa_util.h:143]: This sentence doesn't make sense: `PSA_WANT_ECC_xxx` characterizes a curve, not a curve family. If the combination of `curve` and `bits` is incorrect, that's an absolute statement, not related to a curve.

File: include/mbedtls/psa_util.h
Comment (gilles-peskine-arm) [inline include/mbedtls/psa_util.h:144]: I don't think this sentence is useful at all here. It's covered by the statement above. It just happens that `PSA_WANT_ECC_SECP_K_1` cannot be enabled.

(This statement is useful in the implementation, to explain to maintainers why the case for secp244k1 is missing where all the other curves supported by Mbed TLS are present.)

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I'd still like a few improvements in the documentation.

**Comment by gilles-peskine-arm** [comment]
As long as we aren't enabling secp224k1, if there's some leftover crud, we can merge this PR first and handle secp224k1 in a follow-up. @mpg I'm in meetings all day today and tomorrow, but I'll backport https://github.com/Mbed-TLS/mbedtls/pull/8668 if I get a chance. But if you prefer, feel free to merge it without waiting for the backport.

**Comment by mpg** [comment]
Right, I think I overreacted a bit / lost track of priorities by suggesting we make this depend on 8668. Let's merge this one as soon as it passes CI and is approved, even if we're not fully confident about its handling of secp224k1, and we'll fix it later if needed.

(So, Gilles, no need to rush a backport of 8668.)

**Review by mpg — COMMENTED** [review]
LGTM (I know the CI disagrees, just wanted to record that I've reviewed the last batch of commits).

**Comment by mpg** [comment]
> (So, Gilles, no need to rush a backport of 8668.)

Damn, I was too slow.

**Comment by mpg** [comment]
Now that #8668 is merged, we probably want to rebase that one on development - that way the CI runs will use the correct data as fixed by 8668. (Otherwise there's a risk that the CI appears to pass here, and then fails when we press "merge".)

**Comment by valeriosetti** [comment]
Rebase done. Almost all commits were kept unchanged a part from:
- commits concerning `aymmetric_key_data.py` (which were reverted in the end anyway) due to conflicts. This part was already fixed in #8668
- last commit (https://github.com/Mbed-TLS/mbedtls/pull/8664/commits/500b92b9a9cfffc725fed6d58a1d0fcf0d4ae92b) which removes support for secp224k1 in order to save few more bytes in the code

**Review by mpg — COMMENTED** [review]
Just the note that I've checked the rebase and additional commit and I'm happy with those. Will do a final round of review when the CI is green.

**Comment by valeriosetti** [comment]
> Will do a final (hopefully!) round of review when the CI is green.

I suspect that the OpenCI encountered several errors in the last run as there are many yellow icons there. Can you please restart it?

**Comment by gilles-peskine-arm** [comment]
I've restarted the CI.

**Comment by valeriosetti** [comment]
I forced pushed the last commit because the OpenCI failed again in "Pre-test checks".

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Review by gilles-peskine-arm — APPROVED** [review]
Looks good to me.

The ABI differences reported by the CI are about `mbedtls_ecc_group_of_psa`, which was previously documented as an experimental function, so not actually part of the public ABI, thus a false positive.

**Pull request merged**
